server/graphql/document/src/types/document.rs

```diff
@@ -0,0 +1,112 @@
+use async_graphql::*;
+use async_graphql::{dataloader::DataLoader, Context};
+use chrono::{DateTime, Utc};
+
+use graphql_core::{standard_graphql_error::StandardGraphqlError, ContextExt};
+use repository::Document;
+use service::document::raw_document::RawDocument;
+
+use crate::json_schema_loader::JsonSchemaLoader;
+
+use super::json_schema::JSONSchemaNode;
+
+pub struct DocumentNode {
+    pub document: Document,
+}
+
+#[Object]
+impl DocumentNode {
+    pub async fn id(&self) -> &str {
+        &self.document.id
+    }
+
+    pub async fn name(&self) -> &str {
+        &self.document.name
+    }
+
+    pub async fn parents(&self) -> &[String] {
+        &self.document.parents
+    }
+
+    pub async fn author(&self) -> &str {
+        &self.document.author
+    }
+
+    pub async fn timestamp(&self) -> &DateTime<Utc> {
+        &self.document.timestamp
+    }
+
+    pub async fn r#type(&self) -> &str {
+        &self.document.r#type
+    }
+
+    pub async fn data(&self) -> &serde_json::Value {
+        &self.document.data
+    }
+
+    pub async fn schema(&self, ctx: &Context<'_>) -> Result<Option<JSONSchemaNode>> {
+        Ok(match &self.document.schema_id {
+            Some(schema_id) => {
+                let loader = DataLoader::new(
+                    JsonSchemaLoader {
+                        connection_manager: ctx.get_connection_manager().clone(),
+                    },
+                    async_std::task::spawn,
+                );
+                let schema = loader.load_one(schema_id.clone()).await?.ok_or(
+                    StandardGraphqlError::InternalError(format!(
+                        "Cannot find schema {}",
+                        schema_id
+                    ))
+                    .extend(),
+                )?;
+                Some(JSONSchemaNode { schema })
+            }
+            None => None,
+        })
+    }
+}
+
+#[derive(SimpleObject)]
+pub struct DocumentConnector {
+    pub total_count: u32,
+    pub nodes: Vec<DocumentNode>,
+}
+
+pub struct RawDocumentNode {
+    pub document: RawDocument,
+}
+
+#[Object]
+impl RawDocumentNode {
+    pub async fn name(&self) -> &str {
+        &self.document.name
+    }
+
+    pub async fn parents(&self) -> &[String] {
+        &self.document.parents
+    }
+
+    pub async fn author(&self) -> &str {
+        &self.document.author
+    }
+
+    pub async fn timestamp(&self) -> &DateTime<Utc> {
+        &self.document.timestamp
+    }
+
+    pub async fn r#type(&self) -> &str {
+        &self.document.r#type
+    }
+
+    pub async fn data(&self) -> Result<String> {
+        Ok(serde_json::to_string(&self.document.data).map_err(|e| {
```

Comment:
This seems to be repeated quite a few time, would be good to have a generic serialise/deserialise method that all repositories use (mainly for error mapping to be consistent for all json)

Reply:
I was wondering if we could just unwrap() it. Is there anything that can go wrong serializing a struct? is there invalid json that you can represent in a Rust? I guess custom serialization methods can fail...

I just noticed it should be `fn data(&self) -> Result<serde_json::Value> {}` the other DocumentNode does it correctly...

Reply:
What is the error message you get if you unwrap it ? And given that error message (if it happens in the wild), would we be able replicate it ?

I just realised that this comment was for repository not for graphql

Reply:
you mean when it panics?, think it get caught by actix/graphql and an error is returned...

Reply:
hmm, for to_string case it looks like the only time serde can fail if there are non string keys in map, so this would probably never be the case. My worry about unwraps is an error with not enough info (i think every error should have enough info either logged or return by API. I did a little test with unwrap and deserialisation, I get this api. result:

![Screen Shot 2022-06-01 at 3 50 02 PM](https://user-images.githubusercontent.com/26194949/171324014-55c707cf-cd28-465d-8363-97fe400b4538.png)
![Screen Shot 2022-06-01 at 3 49 30 PM](https://user-images.githubusercontent.com/26194949/171324019-ea231b29-964b-4120-a26d-95ac13c7e13e.png)

And some log does appear in console (btw i should have made erronous result in graphql logging to be error! rather then info!
![Screen Shot 2022-06-01 at 3 50 52 PM](https://user-images.githubusercontent.com/26194949/171324122-4ab8496c-b5b7-4a33-8837-d91b691a55d4.png)

(even built --release version has the line number from lib.rs, that's pretty cool)

Conclusion ?

Can front end handle this kind of response ? (not sure)

Personally I would prefer no unwraps no panics (unless it's during start up)